### DISCUSSION TITLE
Let's get SMART

### DISCUSSION THREAD

**Discussion by jdjkelly** [flexpa/sero#64] [General]
**Big Picture:** Can we build a SMART app + are there tools Sero can offer?

@jamais-vu documenting our previous discussion here, let's use this discussion thread for now. We might create specific issues, but as of right now this is research.

- [ ] Go find the best public guide/tutorial/example you can for creating a SMART on FHIR App and do it - but don't spend more than ~16 hours, using only public resources and your existing knowledge of FHIR
- [ ] And then document different things you find in posts to this thread, your notes along the way

**Comment by jamais-vu** [comment]
Currently, I have a good idea of how SMART apps work and what they can do. The main difficulties I've identified so far are:

- Many examples, especially older ones, use jQuery/AJAX, which I have some experience with but not much.
- Older tutorials such as [Cerner's browser-based app](https://engineering.cerner.com/smart-on-fhir-tutorial/) use [`fhir.js`](https://github.com/FHIR/fhir.js), which is somewhat complex. Fortunately, as of [v2.0](http://docs.smarthealthit.org/client-js/#fhirjs-integration) of the SMART JS Client Library, `fhir.js` is no longer included. Most of its functionality can be obtained via [`client.request`](https://github.com/smart-on-fhir/client-js), which I think is simpler. I don't yet know enough to say whether we should only focus on `client.request` or if we should also consider `fhir.js` for legacy purposes. 
- Most examples lack code comments (this might not be an issue for someone with JS dev experience, but is worth noting).
- I don't have extensive experience with setting up local servers or test environments, so getting different parts to talk to each other is one of the more troublesome parts.

Some useful resources I've found:

- SMART Health IT's [docs](https://docs.smarthealthit.org) are the best starting point for developing SMART apps, and in particular the [SMART JS Client Library docs](http://docs.smarthealthit.org/client-js/) give a succinct overview for JS devs and plenty of code samples.
- SMART Health IT provides an [online SMART app launcher](https://launch.smarthealthit.org), which is super useful and lets developers test a SMART app without having to set up their own test environment.
- The most helpful collection of examples I've found so far is https://github.com/smart-on-fhir/sample-apps. Each app has a README describing how to run it and is simple enough to understand from reading the code. These could make good "building blocks" for referencing in the creation of a more complex app.
- HL7 provides an in-depth guide to the [SMART app launch flow](http://www.hl7.org/fhir/smart-app-launch/), for both apps launched by an EHR and standalone apps. This is the best guide I've found for authorization and access.

Currently I am working on:
1. Set up a HAPI FHIR server.
2. Populate it with Synthea data.
3. Get the super simple SMART app talking to it.
4. Write a SMART app (using `client.request`) that lets the client sign in/out, handles authorization+access tokens, and accesses resources on the FHIR server.

I could do (4) right now by basically gluing together the different examples I've seen so far, but I think doing it from scratch would be instructional and I'd like to see what a more-modern example looks like. 

On the topic of more-modern examples, there are also Typescript+React examples. [This one](https://codesandbox.io/s/fhir-client-typescript-react-spa-4e7rl) in particular is of interest because it also demonstrates use of `client.request`. I'll give a React SMART app a shot after I get a solid understanding of the fundamentals because it could simplify development of these sorts of UIs.

**Comment by jamais-vu** [comment]
I've familiarized myself with HAPI server and set up a local HAPI JPA server following this [starter project](https://github.com/hapifhir/hapi-fhir-jpaserver-starter). 

I chose a [JPA](https://en.wikipedia.org/wiki/Jakarta_Persistence) server over a Plain Server (aka Facade) because the JPA server provides its own relational database and storage/retrieval logic out of the box, which seems well-suited for app development. If I were looking to integrate with existing data sources, I would choose the Plain Server. For more on the different servers see [REST Server Types](https://hapifhir.io/hapi-fhir/docs/server_plain/server_types.html).

The [HAPI FHIR docs](https://hapifhir.io/hapi-fhir/docs/getting_started/introduction.html) are very good, enough that I didn't need to consult any other sources to understand what I was doing. I like that they adequately explain many of the concepts without requiring the reader to have more than basic familiarity with FHIR. 

I only ran into two difficulties. Neither were significant but are worth documenting:
- Running the [HAPI JPA server Docker image](https://github.com/hapifhir/hapi-fhir-jpaserver-starter#running-via-docker-hub) via `docker run -p 8080:8080 hapiproject/hapi:latest` raises:
    >WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested

	This appears to be a common issue running a Docker image on Apple Silicon (eg M1 CPU). Resolved by adding the flag `--platform linux/amd64` to `docker run`.

- Starting the HAPI JPA server itself took several minutes because the Maven build required some downloads. I initially thought this was a silent error related to the Docker issue above, until I also [ran it locally](https://github.com/hapifhir/hapi-fhir-jpaserver-starter#using-jetty) via `mvn jetty:run`. On subsequent runs the server starts in about 15 seconds on my machine. So if it takes a long time at first, that doesn't necessarily mean something is wrong!

Next I'll bulk load Synthea data into the server and connect a simple SMART app.

(Bonus: The server comes with an admin GUI where you can search, query, perform CRUD operations on, and validate FHIR resources.)
<img width="1440" alt="HAPI JPA GUI 01" src="https://user-images.githubusercontent.com/46115232/128179158-0e2873e9-5839-4a0f-af44-c5aa28aa8199.png">
![HAPI JPA GUI 02](https://user-images.githubusercontent.com/46115232/128179639-e1bc4463-b77f-40be-94d9-24337b5df74b.jpg)

**Comment by jamais-vu** [comment]
I can run a very simple browser-based SMART app and request FHIR resources from my server! It returns the searchsets I expect. Feels satisfying.

One file `index.html` (from SMART Health IT's ["Getting started with Browser-based Apps"](https://docs.smarthealthit.org/tutorials/javascript/) tutorial) running on [`http-server`](https://www.npmjs.com/package/http-server):
```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <title>Simple Browser-Based SMART App</title>
        <script src="https://cdn.jsdelivr.net/npm/fhirclient/build/fhir-client.js"></script>
        <script type="text/javascript">
            const client = FHIR.client("http://localhost:8080/fhir");
            client.request("Patient").then(console.log).catch(console.error);
            client.request("Encounter").then(console.log).catch(console.error);
        </script>
    </head>
</html>
```

I'll try something more complex now and see if/where things break.
